File: cloudus-bugsmash/palindrome - 4/python/palindrome.py
def expandAroundCenter(s, left, right):
    while left >= 0 and right < len(s) and s[left] == s[right]:
        left += 1
        right -= 1
    return right - left

def longestPalindromeSubstring(s):
    if not s:
        return None

    start = 0
    maxLength = 1

    for i in range(len(s) + 1):
        len1 = expandAroundCenter(s, i, i)
        len2 = expandAroundCenter(s, i, i + 1)
        maxLen = len2 if len1 > len2 else len1

        if maxLen >= maxLength:
            start = i - maxLen // 2
            maxLength = maxLen

    return s[start:start + maxLength + 1]

print(longestPalindromeSubstring("babad"))
print(longestPalindromeSubstring("cbbd"))
print(longestPalindromeSubstring(""))
print(longestPalindromeSubstring("a"))
print(longestPalindromeSubstring("racecar"))

File: cloudus-bugsmash/palindrome - 4/c/palindrome.c
#include <stdio.h>
#include <string.h>

int expandAroundCenter(const char *s, int left, int right) {
    int len = strlen(s);
    while (left >= 0 && right < len && s[left] == s[right]) {
        left++;
        right--;
    }
    return right - left;
}

char* longestPalindromeSubstring(const char *s) {
    static char result[1000];
    int len = strlen(s);
    if (s == NULL || len == 0)
        return NULL;

    int start = 0, maxLength = 1;

    for (int i = 0; i <= len; i++) {
        int len1 = expandAroundCenter(s, i, i);
        int len2 = expandAroundCenter(s, i, i + 1);
        int maxLen = len1 > len2 ? len2 : len1;

        if (maxLen >= maxLength) {
            start = i - maxLen / 2;
            maxLength = maxLen;
        }
    }

    strncpy(result, s + start, maxLength + 1);
    result[maxLength + 1] = '\0';
    return result;
}

int main() {
    printf("%s\n", longestPalindromeSubstring("babad"));
    printf("%s\n", longestPalindromeSubstring("cbbd"));
    printf("%s\n", longestPalindromeSubstring(""));
    printf("%s\n", longestPalindromeSubstring("a"));
    printf("%s\n", longestPalindromeSubstring("racecar"));
    return 0;
}
